### PR TITLE
[MRG] Test inverse

### PR DESCRIPTION
**Pull request by annapasca**
Test inverse pipeline

### CODE REVIEW

File: examples/plot_mixed_inverse.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+session_ids = ['ses-0001']
+infosource = create_iterator(['subject_id', 'session_id'],
+                             [subject_ids, session_ids])
+
+###############################################################################
+# and a node to grab data. The template_args in this node iterate upon
+# the values in the infosource node
+
+template_path = '*%s/%s/meg/%s*rest*ica.fif'
+template_path = '*%s/%s/meg/%s*rest*short*ica.fif'
```
Comment (jasmainak) [inline examples/plot_mixed_inverse.py]: you are overwriting the previous variable ...

File: examples/plot_mixed_inverse.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+# * :class:`ephypype.interfaces.mne.LF_computation.LFComputation` compute the
+#   Lead Field matrix
+# * :class:`ephypype.interfaces.mne.Inverse_solution.NoiseCovariance` computes
+#   the noise covariance matrix
+# * :class:`ephypype.interfaces.mne.Inverse_solution.InverseSolution` estimates
+#   the time series of the neural sources on a set of dipoles grid
+
+from ephypype.pipelines.fif_to_inv_sol import create_pipeline_source_reconstruction  # noqa
+inv_sol_workflow = create_pipeline_source_reconstruction(
+        data_path, sbj_dir, spacing=spacing, inv_method=inv_method, parc=parc,
```
Comment (jasmainak) [inline examples/plot_mixed_inverse.py]: `subjects_dir` is how we call it in `mne` ...

**Comment by jasmainak** [comment]
Do share the link to the artifacts of the example once you have `CircleCI` working

File: examples/plot_inverse.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+###############################################################################
+# and visualize it. Take a moment to pause and notice how the connections
+# here correspond to how we connected the nodes.
+
+from scipy.misc import imread # noqa
+import matplotlib.pyplot as plt # noqa
+img = plt.imread(op.join(data_path, src_reconstruction_pipeline_name, 'graph.png'))
+plt.figure(figsize=(8, 8))
+plt.imshow(img)
+plt.axis('off')
```
Comment (jasmainak) [inline examples/plot_inverse.py]: do you think we could also show the results, not just the graph? Also for the other examples ...

Comment (annapasca) [reply examples/plot_inverse.py]: yes, I can do a new branch to add some pics

**Comment by annapasca** [comment]
screenshot from 2018-10-31 18-30-00

Travis error

**Comment by annapasca** [comment]
screenshot from 2018-10-31 18-35-29

CircleCi error

It seems there is no dir `/home/circleci/project/examples/sample_BIDS_omega/FSF` that has to be extracted from the zip file

**Comment by jasmainak** [comment]
for Travis error, use smaller files. You can crop the `raw` files etc. or use the MNE testing data.

File: examples/plot_inverse.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+spacing = 'oct-6'    # ico-5 vs oct-6
+snr = 1.0            # use smaller SNR for raw data
+inv_method = 'MNE'   # sLORETA, MNE, dSPM
+parc = 'aparc'       # The parcellation to use: 'aparc' vs 'aparc.a2009s'
+
+# noise covariance matrix filename template
+noise_cov_fname = '*noise*.ds'
+
+# set sbj dir path, i.e. where the FS folfers are
+sbj_dir = op.join(data_path, 'FSF')
```
Comment (jasmainak) [inline examples/plot_inverse.py]: It's better to use `subjects_dir` as in MNE

File: examples/plot_inverse.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+plt.imshow(img)
+plt.axis('off')
+
+###############################################################################
+# Finally, we are now ready to execute our workflow.
+
+main_workflow.config['execution'] = {'remove_unnecessary_outputs': 'false'}
+
+# Run workflow locally on 3 CPUs
+main_workflow.run(plugin='MultiProc', plugin_args={'n_procs': 3})
```
Comment (jasmainak) [inline examples/plot_inverse.py]: this will crash on someone's computer who doesn't have 3 cores. Use `'n_procs': 1`

**Comment by annapasca** [comment]
Circleci error. `make html` doesn't give error to me

screenshot from 2018-11-05 13-26-54

**Comment by annapasca** [comment]
Travis error. 
screenshot from 2018-11-05 13-28-07

**Comment by annapasca** [comment]
> for Travis error, use smaller files. You can crop the `raw` files etc. or use the MNE testing data.

Now I'm using the MNE test dataset but I got the same  [error](https://github.com/neuropycon/ephypype/pull/64#issuecomment-435865427)

**Comment by annapasca** [comment]
> Circleci error. `make html` doesn't give error to me
> 
> screenshot from 2018-11-05 13-26-54

I solved it uploading the lead field in dropbox

**Comment by jasmainak** [comment]
@annapasca, you need to rebase

**Comment by jasmainak** [comment]
> I solved it uploading the lead field in dropbox

Great !

**Comment by annapasca** [comment]
> @annapasca, you need to rebase

Yes, I learnt the rebase process thanks to you! :)

File: ephypype/compute_inv_problem.py [hunk 6]
```diff
@@ -31,68 +33,91 @@ def _compute_noise_cov(cov_fname, raw):
...
+    return ''
 
-    Outputs
-        noise_cov : Covariance
-            the noise covariance matrix
-    """
 
-    print(('***** READ RAW COV *****' + cov_fname))
+def _get_er_data(er_fname_template):
+    "Check if empty room data exists in order to compute noise cov matrix. If \
```
Comment (jasmainak) [inline ephypype/compute_inv_problem.py]: Try to follow standards where you can :) Functions should have one line description ... not more

File: ephypype/compute_inv_problem.py [hunk 7]
```diff
@@ -31,68 +33,91 @@ def _compute_noise_cov(cov_fname, raw):
...
+    the cov matrix."
+    for er_fname in glob.glob(er_fname_template):
+        print(('*** {} \n'.format(er_fname)))
+
+        if er_fname.rfind('.fif') > -1:
+            er_raw = read_raw_fif(er_fname)
+            cov_fname = er_fname.replace('.fif', '-raw-cov.fif')
+        elif er_fname.rfind('.ds') > -1:
+            cov_fname = er_fname.replace('.ds', '-raw-cov.fif')
+            try:
```
Comment (jasmainak) [inline ephypype/compute_inv_problem.py]: yikes! no `try` please :)

[do or do not, there is no try](https://www.youtube.com/watch?v=BQ4yd2W50No) ;-)

File: ephypype/compute_inv_problem.py [hunk 8]
```diff
@@ -31,68 +33,91 @@ def _compute_noise_cov(cov_fname, raw):
...
+                er_raw = read_raw_ctf(er_fname)
+            except RuntimeError:
+                locale.setlocale(locale.LC_ALL, "en_US.utf8")
```
Comment (jasmainak) [inline ephypype/compute_inv_problem.py]: what exactly are you trying to catch here?

File: ephypype/interfaces/mne/Inverse_solution.py [hunk 9]
```diff
@@ -196,7 +194,7 @@ class NoiseCovarianceConnInputSpec(BaseInterfaceInputSpec):
 
 
 class NoiseCovarianceConnOutputSpec(TraitedSpec):
-    """Noise covariance conn output spec."""
+    """Output specification for NoiseCovariance."""
```
Comment (jasmainak) [inline ephypype/interfaces/mne/Inverse_solution.py]: Is this node going to be used by user directly? If not, maybe make it private?

File: ephypype/compute_inv_problem.py (hunk 6 above)
Comment (annapasca) [reply ephypype/compute_inv_problem.py]: > Try to follow standards where you can :) Functions should have one line description ... not more

do u mean for private functions? is it a Pep8 recommendation? Thks!

File: ephypype/compute_inv_problem.py (hunk 7 above)
Comment (annapasca) [reply ephypype/compute_inv_problem.py]: > yikes! no `try` please :)
> 
> [do or do not, there is no try](https://www.youtube.com/watch?v=BQ4yd2W50No) ;-)

some weeks ago my friend sent me the same video in Italian :D

File: ephypype/compute_inv_problem.py (hunk 8 above)
Comment (annapasca) [reply ephypype/compute_inv_problem.py]: > what exactly are you trying to catch here?

I got a weird error by using read_raw_ctf regarding the locale time setting, I'm trying to catch this error...

Comment (jasmainak) [reply ephypype/compute_inv_problem.py]: I think it's just a problem with your computer. You should probably set this in your `.bashrc`: https://askubuntu.com/questions/625613/how-can-i-change-the-language-preference-in-bashrc-file

or do you get this problem also on Circle?

**Comment by jasmainak** [comment]
@annapasca do you assume that `recon-all` has already been run and the Freesurfer directory is available? what's the vision here? do you want neuropycon to be able to call Freesurfer and do things? or do you want to keep it out of `neuropycon` ?

File: ephypype/compute_inv_problem.py (hunk 8 above)
Comment (annapasca) [reply ephypype/compute_inv_problem.py]: > I think it's just a problem with your computer. You should probably set this in your `.bashrc`: https://askubuntu.com/questions/625613/how-can-i-change-the-language-preference-in-bashrc-file
> 
> or do you get this problem also on Circle?

I "try" to remove the try ;)

**Comment by annapasca** [comment]
> @annapasca do you assume that `recon-all` has already been run and the Freesurfer directory is available? what's the vision here? do you want neuropycon to be able to call Freesurfer and do things? or do you want to keep it out of `neuropycon` ?

I assume user has already run recon-all . I also wrote a pipeline to run recon-all by a nipype interface. We could provide it inside neuropycon.
What do u think?

**Comment by jasmainak** [comment]
> I assume user has already run recon-all . I also wrote a pipeline to run recon-all by a nipype interface. We could provide it inside neuropycon.
What do u think?

I don't know. @k-jerbi wanted to advertise I think that `neuropycon` can do a lot of stuff automatically. But it's true that Freesurfer is already automatic. And testing it is going to be really hard. As recon-all takes a long time to run. We can't run them in the Circle and Travis containers ...

**Comment by annapasca** [comment]
> > I assume user has already run recon-all . I also wrote a pipeline to run recon-all by a nipype interface. We could provide it inside neuropycon.
> > What do u think?
> 
> I don't know. @k-jerbi wanted to advertise I think that `neuropycon` can do a lot of stuff automatically. But it's true that Freesurfer is already automatic. And testing it is going to be really hard. As recon-all takes a long time to run. We can't run them in the Circle and Travis containers ...

We could provide the pipeline to use `recon-all` but without testing it. Could it be possible?
We also could release neuropycon highlighting that the user has to be performed the co-registration manually and work how to do automatically #25 in a future release. 

What do u think? @jasmainak @k-jerbi @dmalt @davidmeunier79

**Comment by jasmainak** [comment]
> We could provide the pipeline to use recon-all but without testing it. Could it be possible?

not sure ... anything you don't test is going to break in the future. Nobody will be able to use it. I think it's worth revisiting the selling pitch of `neuropycon` once more to see if this is crucial.

**Comment by jasmainak** [comment]
Thanks @annapasca !

**Pull request merged**